{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7e17c9d3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Invalid license file syntax.\n",
      "Feature:       PYMOL_MAIN\n",
      "License path:  /home/sirius/.pymol/license.lic:\n",
      "FlexNet Licensing error:-2,413\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import biotite\n",
    "from biotite.structure.residues import get_residues\n",
    "from biotite.structure import get_chains\n",
    "from biotite.sequence import ProteinSequence\n",
    "import biotite.structure as struc\n",
    "import biotite.structure.io as strucio\n",
    "import glob,re\n",
    "\n",
    "import os,sys,time,json,fnmatch,shutil,re,subprocess,multiprocessing\n",
    "import numpy as np\n",
    "import torch\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning, module=\"biotite\")\n",
    "\n",
    "from pymol import cmd\n",
    "import copy\n",
    "\n",
    "def single_chain_rmsd(template, temp_pdbfile,positin=None):\n",
    "    if isinstance(template,str):\n",
    "        template = strucio.load_structure(template,model=1)\n",
    "        template = template[template.atom_name == \"CA\"]\n",
    "    if isinstance(temp_pdbfile,str):\n",
    "        temp_pdbfile = strucio.load_structure(temp_pdbfile,model=1)\n",
    "        temp_pdbfile = temp_pdbfile[temp_pdbfile.atom_name == \"CA\"]\n",
    "    if positin:\n",
    "        res_id = template.res_id\n",
    "        res_id_mask = np.isin(res_id,positin)\n",
    "        template = template[res_id_mask]\n",
    "        temp_pdbfile = temp_pdbfile[res_id_mask]\n",
    "    superimposed, _ = struc.superimpose(template, temp_pdbfile)\n",
    "    motif_rmsd = struc.rmsd(template, superimposed)\n",
    "    return motif_rmsd\n",
    "\n",
    "def usalign_wrapper(template, temp_pdbfile, force_alignment=None):\n",
    "    if force_alignment == None:\n",
    "        p = subprocess.Popen(f'/home/sirius/Desktop/USalign {template} {temp_pdbfile} -mm 1 -ter 0 | grep -E \"RMSD|TM-score=\" ', stdout=subprocess.PIPE, shell=True)\n",
    "    else:\n",
    "        p = subprocess.Popen(f'/home/sirius/Desktop/USalign {template} {temp_pdbfile} -I {force_alignment} -mm 1 -ter 0| grep -E \"RMSD|TM-score=\" ', stdout=subprocess.PIPE, shell=True)\n",
    "    output, __ = p.communicate()\n",
    "    output = output.decode('utf-8')\n",
    "    tm_rmsd = float(re.search('RMSD=\\s*(\\d+\\.\\d+)', output).group(1))\n",
    "    tm_score = float(re.search('TM-score=\\s*(\\d+\\.\\d+)', output).group(1))\n",
    "    return tm_rmsd, tm_score\n",
    "\n",
    "def tmalign_wrapper(template, temp_pdbfile, force_alignment=None):\n",
    "    if force_alignment == None:\n",
    "        p = subprocess.Popen(f'/home/sirius/Desktop/TMalign {template} {temp_pdbfile} | grep -E \"RMSD|TM-score=\" ', stdout=subprocess.PIPE, shell=True)\n",
    "    else:\n",
    "        p = subprocess.Popen(f'/home/sirius/Desktop/TMalign {template} {temp_pdbfile} -I {force_alignment} | grep -E \"RMSD|TM-score=\" ', stdout=subprocess.PIPE, shell=True)\n",
    "    output, __ = p.communicate()\n",
    "    tm_rmsd  = float(str(output)[:-3].split(\"RMSD=\")[-1].split(\",\")[0] )\n",
    "    tm_score = float(str(output)[:-3].split(\"TM-score=\")[-1].split(\"(if\")[0] )\n",
    "    return tm_rmsd, tm_score\n",
    "\n",
    "def extract_plddt(protein,chain_id=None):\n",
    "    if isinstance(protein,str):\n",
    "        # model = 1 to load a AtomArray object\n",
    "        # extra_fields to load the b_factor column\n",
    "        atom_array = strucio.load_structure(protein,model=1,extra_fields=[\"b_factor\"])\n",
    "    elif isinstance(protein, struc.AtomArrayStack):\n",
    "        atom_array = protein[0]\n",
    "    elif isinstance(protein, struc.AtomArray):\n",
    "        atom_array = protein\n",
    "\n",
    "    # add multiple chain sequence subtract function\n",
    "    all_chains = get_chains(atom_array)\n",
    "    if len(all_chains) == 0:\n",
    "        raise ValueError('No chains found in the input file.')\n",
    "    if chain_id is None:\n",
    "        chain_ids = all_chains\n",
    "    elif isinstance(chain_id, list):\n",
    "        chain_ids = chain_id\n",
    "    else:\n",
    "        chain_ids = [chain_id] \n",
    "    for chain in chain_ids:\n",
    "        if chain not in all_chains:\n",
    "            raise ValueError(f'Chain {chain} not found in input file')\n",
    "    chain_filter = [a.chain_id in chain_ids for a in atom_array]\n",
    "    atom_array = atom_array[chain_filter]\n",
    "\n",
    "    # mask canonical aa \n",
    "    aa_mask = struc.filter_amino_acids(atom_array) # higher version of biotite\n",
    "    atom_array = atom_array[aa_mask]\n",
    "\n",
    "    # ca atom only\n",
    "    atom_array = atom_array[atom_array.atom_name == \"CA\"]\n",
    "\n",
    "    plddt = np.array([i.b_factor for i in atom_array])\n",
    "\n",
    "    return np.mean(plddt)\n",
    "\n",
    "# ***********************************************************************************************************\n",
    "\n",
    "def chain_pair_pae(pae, chain_number):\n",
    "    # 判断pae的shape[0]可以被chain_number整除\n",
    "    assert pae.shape[0] % chain_number == 0\n",
    "    chain_length = pae.shape[0] // chain_number\n",
    "    interaction_mask = np.ones_like(pae)\n",
    "    for i in range(chain_number):\n",
    "        interaction_mask[i*chain_length:(i+1)*chain_length, i*chain_length:(i+1)*chain_length] = 0.0\n",
    "    return np.sum(pae * interaction_mask) / np.sum(interaction_mask)\n",
    "\n",
    "def get_chain_number_from_pdb(pdb_file):\n",
    "    atom_array = strucio.load_structure(pdb_file,model=1)\n",
    "    all_chains = get_chains(atom_array)\n",
    "    return len(all_chains)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def pymol_align(folder_pdb,save_path):\n",
    "    # only for ProteinMPNN lmpnn and NeMO_v3, not suitable for NeMO_v2\n",
    "    cmd.reinitialize()\n",
    "    cmd.do(\"run /home/sirius/Desktop/tmalign.py\")\n",
    "    all_file = os.listdir(folder_pdb)\n",
    "    # Sort files based on the number extracted from the file name\n",
    "\n",
    "    all_file = sorted(\n",
    "        os.listdir(folder_pdb),\n",
    "        key=lambda x: (\n",
    "            # 优先按 T_ 的数字排序（T_后面的数字，比如 T_0.1 转换为 0.1）\n",
    "            float(re.search(r\"T_(\\d\\.\\d)\", x).group(1)) if \"T_\" in x else float(\"inf\"),\n",
    "            # 再按 sample 的数字排序（sample 后面的数字）\n",
    "            int(re.search(r\"sample_(\\d+)_\", x).group(1)) if \"sample_\" in x else -1\n",
    "        )\n",
    "    )\n",
    "\n",
    "    for file in all_file:\n",
    "        if \"seed\" in file:\n",
    "            name = \"native\"\n",
    "        else:\n",
    "            sample_pattern = r\"sample_(\\d+)_\"\n",
    "            match = re.search(sample_pattern, file)\n",
    "            name_id = match.group(1)\n",
    "            # T_0.1__sample_67__score_0.4917__global_score_0.4917__seq_recovery_0.4400\n",
    "            temperature_pattern = r\"T_(\\d+\\.\\d+)\"\n",
    "            temperature_match = re.search(temperature_pattern, file)\n",
    "            temperature = temperature_match.group(1) \n",
    "            name = f\"{temperature}_{name_id}\"\n",
    "        cmd.load(os.path.join(folder_pdb,file), name)\n",
    "\n",
    "    for obj in cmd.get_object_list():\n",
    "        cmd.do(f\"tmalign {obj}, {cmd.get_object_list()[0]}\")\n",
    "    cmd.set('grid_mode', 1)\n",
    "    cmd.zoom()\n",
    "    cmd.remove('solvent')\n",
    "    cmd.do(\"@/home/sirius/Desktop/xcf.pml\")\n",
    "    cmd.save(save_path)\n",
    "# ***********************************************************************************************************\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def extract_seq(protein, chain_id=None):\n",
    "    if isinstance(protein, str):\n",
    "        atom_array = strucio.load_structure(protein, model=1)\n",
    "    elif isinstance(protein, struc.AtomArrayStack):\n",
    "        atom_array = protein[0]\n",
    "    elif isinstance(protein, struc.AtomArray):\n",
    "        atom_array = protein\n",
    "    # aa_mask = struc.filter_canonical_amino_acids(atom_array)\n",
    "    # atom_array = atom_array[aa_mask]\n",
    "    all_chains = get_chains(atom_array)\n",
    "    if len(all_chains) == 0:\n",
    "        raise ValueError('No chains found in the input file.')\n",
    "    if chain_id is None:\n",
    "        chain_ids = all_chains\n",
    "    elif isinstance(chain_id, list):\n",
    "        chain_ids = chain_id\n",
    "    else:\n",
    "        chain_ids = [chain_id]\n",
    "    for chain in chain_ids:\n",
    "        if chain not in all_chains:\n",
    "            raise ValueError(f'Chain {chain} not found in input file')\n",
    "    chain_filter = [a.chain_id in chain_ids for a in atom_array]\n",
    "    atom_array = atom_array[chain_filter]\n",
    "    # mask canonical aa\n",
    "    aa_mask = struc.filter_amino_acids(atom_array)\n",
    "    atom_array = atom_array[aa_mask]\n",
    "    residue_identities = get_residues(atom_array)[1]\n",
    "    seq = ''.join([ProteinSequence.convert_letter_3to1(r)\n",
    "                  for r in residue_identities])\n",
    "    return seq\n",
    "\n",
    "def usalign_wrapper(template, temp_pdbfile, force_alignment=None):\n",
    "    if force_alignment == None:\n",
    "        p = subprocess.Popen(f'/home/sirius/Desktop/USalign {template} {temp_pdbfile} -mm 1 -ter 0 | grep -E \"RMSD|TM-score=\" ', stdout=subprocess.PIPE, shell=True)\n",
    "    else:\n",
    "        p = subprocess.Popen(f'/home/sirius/Desktop/USalign {template} {temp_pdbfile} -I {force_alignment} -mm 1 -ter 0| grep -E \"RMSD|TM-score=\" ', stdout=subprocess.PIPE, shell=True)\n",
    "    output, __ = p.communicate()\n",
    "    output = output.decode('utf-8')\n",
    "    tm_rmsd = float(re.search('RMSD=\\s*(\\d+\\.\\d+)', output).group(1))\n",
    "    tm_score = float(re.search('TM-score=\\s*(\\d+\\.\\d+)', output).group(1))\n",
    "    return tm_rmsd, tm_score\n",
    "\n",
    "def mmalign_wrapper(template, temp_pdbfile, force_alignment=None):\n",
    "    if force_alignment == None:\n",
    "        p = subprocess.Popen(f'/home/sirius/Desktop/MMalign {template} {temp_pdbfile} | grep -E \"RMSD|TM-score=\" ', stdout=subprocess.PIPE, shell=True)\n",
    "    else:\n",
    "        p = subprocess.Popen(f'/home/sirius/Desktop/MMalign {template} {temp_pdbfile} -I {force_alignment} | grep -E \"RMSD|TM-score=\" ', stdout=subprocess.PIPE, shell=True)\n",
    "    output, __ = p.communicate()\n",
    "    output = output.decode('utf-8')  # Decode the bytes to a string\n",
    "    tm_rmsd  = float(str(output)[:-3].split(\"RMSD=\")[-1].split(\",\")[0] )\n",
    "    tm_score = float(str(output)[:-3].split(\"TM-score=\")[-1].split(\"(normalized\")[0] )\n",
    "    aligned_length_line = [line for line in output.split('\\n') if 'Aligned length=' in line][0]\n",
    "    aligned_length = int(aligned_length_line.split(\"Aligned length=\")[-1].split(\",\")[0].strip())\n",
    "    seq_id = float(aligned_length_line.split(\"Seq_ID=n_identical/n_aligned=\")[-1].split()[0].strip())\n",
    "    return tm_rmsd, tm_score, aligned_length, seq_id\n",
    "\n",
    "def load_mpnn_seq(fasta=None,folder=None):\n",
    "    \"\"\"\n",
    "    Load Protienmpnn / nemo style fasta \n",
    "    1. give a fasta return all the sequence\n",
    "    2. give a folder return all the fasta in the folder\n",
    "\n",
    "    return a pandas dataframe with columns: fasta_name, T, sample, score, global_score, seq_recovery, seq\n",
    "    \"\"\"\n",
    "    all_seqs = []\n",
    "    exists_seq = set()\n",
    "    repeat_label = False\n",
    "    assert (fasta is not None) != (folder is not None), \"only one input is allowed!\"\n",
    "    if fasta is not None:\n",
    "        fasta_name = \".\".join(fasta.split(\"/\")[-1].split(\".\")[:-1])\n",
    "        with open(fasta, \"r\") as f:\n",
    "            all_data = [i.strip() for i in f.readlines()]\n",
    "            header = [i[1:] for i in all_data if i.startswith(\">\")]\n",
    "            seq = [i for i in all_data if not i.startswith(\">\")]\n",
    "            # remove the first sequence which is the native sequence\n",
    "            header = header[1:]\n",
    "            seq = seq[1:]\n",
    "            for h,s in zip(header,seq):\n",
    "                if s not in exists_seq:\n",
    "                    exists_seq.add(s)\n",
    "                    # T=0.1, sample=1, score=0.6639, global_score=0.8922, seq_recovery=0.2500\n",
    "                    # extract values from header line (h)\n",
    "                    matches = re.findall(r'(T|sample|score|global_score|seq_recovery)=([\\d\\.]+)', h)\n",
    "                    values = {k: float(v) if '.' in v else int(v) for k,v in matches}\n",
    "                    T = values.get('T')\n",
    "                    sample = values.get('sample')\n",
    "                    score = values.get('score')\n",
    "                    global_score = values.get('global_score')\n",
    "                    seq_recovery = values.get('seq_recovery')\n",
    "                    all_seqs.append([fasta_name,float(T),int(sample),float(score),float(global_score),float(seq_recovery),s])\n",
    "                else:\n",
    "                    repeat_label = True\n",
    "    if folder is not None:\n",
    "        for fasta in os.listdir(folder):\n",
    "            if fasta.endswith(\".fa\"):\n",
    "                fasta_name = fasta.replace(\".fa\", \"\")\n",
    "            elif fasta.endswith(\".fasta\"):\n",
    "                fasta_name = fasta.replace(\".fasta\", \"\")\n",
    "            else:\n",
    "                continue\n",
    "            with open(os.path.join(folder,fasta), \"r\") as f:\n",
    "                all_data = [i.strip() for i in f.readlines()]\n",
    "                header = [i[1:] for i in all_data if i.startswith(\">\")]\n",
    "                seq = [i for i in all_data if not i.startswith(\">\")]\n",
    "                header = header[1:]\n",
    "                seq = seq[1:]\n",
    "                for h,s in zip(header,seq):\n",
    "                    if s not in exists_seq:\n",
    "                        exists_seq.add(s)\n",
    "                        matches = re.findall(r'(T|sample|score|global_score|seq_recovery)=([\\d\\.]+)', h)\n",
    "                        values = {k: float(v) if '.' in v else int(v) for k,v in matches}\n",
    "                        T = values.get('T')\n",
    "                        sample = values.get('sample')\n",
    "                        score = values.get('score')\n",
    "                        global_score = values.get('global_score')\n",
    "                        seq_recovery = values.get('seq_recovery')\n",
    "                        all_seqs.append([fasta_name,float(T),int(sample),float(score),float(global_score),float(seq_recovery),s])\n",
    "                    else:\n",
    "                        repeat_label = True\n",
    "    if repeat_label:\n",
    "        print(f\"Warning: {fasta_name} has repeat label!\")\n",
    "    all_seqs = pd.DataFrame(all_seqs,columns=[\"fasta_name\",\"T\",\"sample\",\"score\",\"global_score\",\"seq_recovery\",\"seq\"])\n",
    "    return all_seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a98af93",
   "metadata": {},
   "outputs": [],
   "source": [
    "import py2Dmol\n",
    "\n",
    "viewer = py2Dmol.view()\n",
    "\n",
    "viewer.add_pdb(\"bioinformatics/7z09_relaxed.pdb\")\n",
    "viewer.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ecf0413c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dl",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
